src/reconstruction/community.jl

```diff
@@ -0,0 +1,409 @@
+            I[kstart] = n_metabolites_metabolic + n
+            J[kstart] = ex_rxn + reaction_offset[i]
+            V[kstart] = -stoichs[i][ex_met, ex_rxn] # ex is normally negative, make positive
+            kstart += 1
+        end
+    end
+
+    # # build diagonal environmental exchanges
+    for i = 1:length(exchange_rxn_ids)
+        I[kstart] = n_metabolites_metabolic + i
+        J[kstart] = n_reactions_metabolic + i
+        V[kstart] = -1.0
+        kstart += 1
+    end
+
+    if add_biomass_objective
+        n_before_biomass_row = n_metabolites_metabolic + length(exchange_met_ids)
+        for i = 1:length(models)
+            biomass_ind = first(indexin([biomass_ids[i]], reactions(models[i])))
+            I[kstart] = i + n_before_biomass_row
+            J[kstart] = biomass_ind + reaction_offset[i]
+            V[kstart] = 1.0
+            kstart += 1
+        end
+    end
+
+    S = sparse(
+        I[1:kstart-1],
+        J[1:kstart-1],
+        V[1:kstart-1],
+        n_metabolites_total,
+        n_reactions_total,
+    ) # could be that some microbes don't have all the exchanges, hence kstart-1
+
+    reaction_cumsum = cumsum(reaction_lengths)
+    metabolite_cumsum = cumsum(metabolite_lengths)
+    for i = 1:length(models)
+        species = isempty(species_names) ? "species_$(i)" : species_names[i]
+        tlbs, tubs = bounds(models[i])
+        lbs[reaction_offset[i]+1:reaction_cumsum[i]] .= tlbs
+        ubs[reaction_offset[i]+1:reaction_cumsum[i]] .= tubs
+        rxns[reaction_offset[i]+1:reaction_cumsum[i]] =
+            "$(species)_" .* reactions(models[i])
+        mets[metabolite_offset[i]+1:metabolite_cumsum[i]] =
+            "$(species)_" .* metabolites(models[i])
+    end
+    mets[metabolite_cumsum[end]+1:metabolite_cumsum[end]+length(exchange_met_ids)] .=
+        exchange_met_ids
+    rxns[reaction_cumsum[end]+1:reaction_cumsum[end]+length(exchange_rxn_ids)] .=
+        exchange_rxn_ids
+
+    if add_biomass_objective
+        rxns[end] = "community_biomass"
+        for i = 1:length(models)
+            species = isempty(species_names) ? "species_$(i)" : species_names[i]
+            mets[end-length(biomass_ids)+i] = "$(species)_" .* biomass_ids[i]
+        end
+    end
+
+    return CoreModel(S, spzeros(size(S, 1)), spzeros(size(S, 2)), lbs, ubs, rxns, mets)
+end
+
+"""
+    all_boundaries(model::M) where {M<:MetabolicModel}
+
+Return `boundary_rxns` and `boundary_mets` given `model`. Boundary reactions
+are reactions with a single stoichiometric coefficient in a column in the
+stoichiometric matrix, and boundary metabolites are the corresponding row/metabolite
+for that column.
+"""
+function all_boundaries(model::M) where {M<:MetabolicModel}
+    boundary_mets = String[]
+    boundary_rxns = String[]
+    S = stoichiometry(model)
+    rxns = reactions(model)
+    mets = metabolites(model)
+    for i = 1:size(S, 2)
+        j, b = findnz(S[:, i])
```

Comment:
use `nnz` or something similar to avoid spawning the long vectors

Reply:
done!